### PR TITLE
OkHttp: Fix NetworkOnMainThreadException while cancellation

### PR DESCRIPTION
**Pull request by yongjhih**
See square/okhttp#869 also.

### CODE REVIEW

File: imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetchProducer.java [hunk 1]
```diff
@@ -83,7 +85,17 @@ protected void fetchImage(final NfpRequestState requestState) {
           new BaseProducerContextCallbacks() {
             @Override
             public void onCancellationRequested() {
-              call.cancel();
+              // Avoid NetowkrOnMainThreadException
+              if (Looper.myLooper() != Looper.getMainLooper()) {
+                call.cancel();
+              } else {
+                new AsyncTask<Call, Void, Void>() {
```
Comment (jasta) [inline imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetchProducer.java]: `AsyncTask` is overkill here.  Specifically, `AsyncTask` is needed only when you require a response in some fashion from the background job on a Handler-based thread (typically the main thread).  In this case, you have fire-and-forget semantics, which a simple Thread or Executor will do much better.

I'm not as familiar with Fresco's code, but I suspect there is a standard internal Executor (or something just like it) that can be used for this purpose as well.

File: imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetchProducer.java [hunk 2]
```diff
@@ -83,7 +85,17 @@ protected void fetchImage(final NfpRequestState requestState) {
           new BaseProducerContextCallbacks() {
             @Override
             public void onCancellationRequested() {
-              call.cancel();
+              // Avoid NetowkrOnMainThreadException
```
Comment (jasta) [inline imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetchProducer.java]: Typo here and in your commit message.

Comment (yongjhih) [inline imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetchProducer.java]: Thanks for correction

**Comment by tyronen** [comment]
Hi Andrew, thanks for doing this.

I see here though that, if the Executor is not supplied, you're creating a whole new thread pool just for cancellations. 

OkHttpClient itself doesn't seem to expose its thread pool, nor offer an async cancel method which would be a proper solution. 

A workaround might be to use [ExecutorSupplier.forBackground](http://frescolib.org/javadoc/reference/com/facebook/imagepipeline/core/ExecutorSupplier.html#forBackground\(\)) method.

What you can do is make an instance of this an optional argument to OkHttpImagePipelineConfigFactory.newBuilder. If it isn't supplied, use DefaultExecutorSupplier. Then pass it to OkHttpNetworkFetchProducer, and before returning call ImagePipelineConfig.Builder.setExecutorSupplier.

**Comment by jasta** [comment]
I believe okhttp does expose its executor via
client.getDispatcher().getExecutorService().  That's probably your easiest
option to fix this.

**Comment by yongjhih** [comment]
Thanks @tyronen and @jasta. Indeed, I thought options of ExecutorSupplier.forBackground() and client.getDispatcher().getExecutorService() but I didn't finish it.

First, it's complicated to pass executorSupplier to okHttpNetworkFetchProducer.
OkHttpImagePipelineConfigFactory provides an imagePipelineConfigBuilder that already setup OkHttpNetworkFetchProducer. How to pass executorSupplier to okHttpNetworkFetchProducer when ImagePipelineConfigBuilder.setExecutorSupplier(executorSupplier)? Inherit ImagePipelineConfigBuilder?

I'm not as familiar with okhttpclient.getDispatcher().getExecutorService() so I have not tried yet.

**Comment by tyronen** [comment]
client.getDispatcher is a better solution anyway. The network library's threads should be used for network operations.

**Comment by yongjhih** [comment]
Thanks @tyronen, I tested it, it works. So I updated PR.

**Comment by tyronen** [comment]
Hi sorry for the delay in getting back to you, it was Easter holiday. We have some other changes in head that you'll need to rebase off of.

**Comment by yongjhih** [comment]
@tyronen, It's ok. I rebased.

**Comment by tyronen** [comment]
LGTM

File: imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetcher.java [hunk 3]
```diff
@@ -37,11 +39,14 @@
 
   private final OkHttpClient mOkHttpClient;
 
+  private Executor mCancellationExecutor;
```
Comment (DHuckaby) [inline imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetcher.java]: This field should be final

Comment (yongjhih) [inline imagepipeline-backends/imagepipeline-okhttp/src/main/java/com/facebook/imagepipeline/backends/okhttp/OkHttpNetworkFetcher.java]: It's rebasing mistake, Thanks.

**Pull request closed without merging**